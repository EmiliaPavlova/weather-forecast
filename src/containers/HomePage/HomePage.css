.wrapper {
  background: #FCFEFC;
}

.loading {
  display: flex;
  justify-content: space-around;
}