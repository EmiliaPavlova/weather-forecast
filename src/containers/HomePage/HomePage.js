import React, { useState, useEffect } from 'react';
import spinner from '../../assets/loading.gif';
import Header from '../../components/Header/Header';
import Search from '../../components/Search/Search';
import ForecastBlock from '../ForecastBlock/ForecastBlock';
import './HomePage.css';

const apiKey = '93af8c1ec5f32e1de2beb62d0b681c7a'; // gmail
// const apiKey = '5764b30726c51254cba63dbbfc6ea136'; // yahoo
const defaultCity = 'Sofia,BG';

export default function HomePage() {
  const [coordinates, setCoordinates] = useState();
  const [forecast, setForecast] = useState();
  const [searchData, setSearchData] = useState();
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();
  const [location, setLocation] = useState();

  useEffect(() => {
    const locationData = `http://api.openweathermap.org/geo/1.0/direct?q=${location || defaultCity}&limit=5&appid=${apiKey}`;
    fetch(locationData)
      .then(response => response.json())
      .then(data => {
        console.log(data)
        data.length > 1
          ? setSearchData(data)
          : setCoordinates(data[0]);
      })
      .catch(error => setError(error))
  }, [location])

  useEffect(() => {
    coordinates && fetch(`http://api.openweathermap.org/data/2.5/forecast?lat=${coordinates.lat}&lon=${coordinates.lon}&appid=${apiKey}`)
      .then(response => response.json())
      .then(data => {
        console.log(data)
        setForecast(data.list);
        setLoading(false);
      })
      .catch(error => {
        setError(error);
        setLoading(false);
      })
  }, [coordinates])

  const title = `5 day weather forecast for ${(coordinates && coordinates.name)}`;

  const onSearch = (string) => setLocation(string)

  return (
    <div className='wrapper'>
      <Header title={title} />
      <Search onSearch={onSearch} onSelectCity={(city) => setCoordinates(city)} results={searchData} />
      <div className=''>
        {loading
          ? <div className='loading'><img src={spinner} alt='loading' /></div>
          : coordinates
            ? <ForecastBlock forecast={forecast}/>
            : error
              ? <div>An error occured</div>
              : <div>No data found</div>
        }
      </div>
    </div>
  );
}