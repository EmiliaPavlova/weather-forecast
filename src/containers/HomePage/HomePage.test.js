import React from 'react';
import {
  render,
  screen,
  cleanup,
  fireEvent
} from '@testing-library/react';
import { shallow } from 'enzyme';
import HomePage from './HomePage';

describe('HomePage component', () => {
  afterEach(cleanup);

  it('matches the snapshot', () => {
    const wrapper = shallow(<HomePage />);
    expect(wrapper).toMatchSnapshot();
  })

  it('calls getLocation on button click', async () => {
    const mockGeolocation = {
      getCurrentPosition: jest.fn()
        .mockImplementationOnce((success) => Promise.resolve(success({
          coords: { latitude: 51.1, longitude: 45.3 }
        })))
    };
    global.navigator.geolocation = mockGeolocation;
    render(<HomePage />);
    await fireEvent.click(screen.getByText('Get current location'));
    expect(mockGeolocation.getCurrentPosition).toHaveBeenCalled();
  });

  it('gets data for given city', async () => {
    const getCoordinationsMock = jest.spyOn(global, 'fetch').mockResolvedValue({
      json: jest.fn().mockResolvedValue([{ name: 'Sofia', country: 'BG', lat: 42.6978634, lon: 23.3221789 }])
    })
    await render(<HomePage />);
    expect(getCoordinationsMock).toHaveBeenCalled();
  });
})