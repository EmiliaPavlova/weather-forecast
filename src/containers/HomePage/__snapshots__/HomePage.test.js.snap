// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage component matches the snapshot 1`] = `
<div
  className="wrapper"
>
  <Header
    title="5 day weather forecast for undefined"
  />
  <div
    className="current_location_wrapper"
  >
    <button
      className="current_location"
      onClick={[Function]}
    >
      Get current location
    </button>
  </div>
  <Search
    onSearch={[Function]}
    onSelectCity={[Function]}
  />
  <div
    className="loading"
  >
    <img
      alt="loading"
      src="loading.gif"
    />
  </div>
</div>
`;
